### PR TITLE
Improving testability and intuitivity of hooks 

### PR DESCRIPTION
**Pull request by maurei**
We found that when implementing resource hooks in our application, it wasn't very easy to unit test them. This was because 1) the resource helpers `AffectedRelationships`, `AffectedResources`, and `AffectedResourceDiff` were hard/impossible to mock/instantiate for unit testing, and because 2) these types don't give an intuitive feeling of the data structure they represent.

This PR fixes this by introducing following changes

* made the constructors of the affected resource/relationship helper classes public so they can be instantiated
* these public constructors now have clear dependencies
* got rid of `RelationshipProxy` where the use of `RelationshipAttribute` suffices; this should make it easier to instantiate `AffectedRelationships`.
* `AffectedRelationships`, `AffectedResources` and `AffectedResourceDiff` have been renamed to `RelationshipsDictionary`, `EntityHashSet` and `EntityDiffs` to make clear that (for the first two) you're actually just dealing with a dictionary and a hashset.

### CODE REVIEW

File: src/JsonApiDotNetCore/Hooks/IResourceHookContainer.cs [hunk 1]
```diff
@@ -194,7 +140,69 @@ public interface IBeforeHooks<TEntity> where TEntity : class, IIdentifiable
         /// <returns>The transformed set of ids</returns>
         /// <param name="resourcesByRelationship">A helper that groups the entities by the affected relationship</param>
         /// <param name="pipeline">An enum indicating from where the hook was triggered.</param>
-        void BeforeImplicitUpdateRelationship(IAffectedRelationships<TEntity> resourcesByRelationship, ResourcePipeline pipeline);
+        void BeforeImplicitUpdateRelationship(IRelationshipsDictionary<TEntity> resourcesByRelationship, ResourcePipeline pipeline);
+    }
+
+
+
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Hooks/IResourceHookContainer.cs]: empty spaaace

File: src/JsonApiDotNetCore/Internal/TypeHelper.cs [hunk 2]
```diff
@@ -1,11 +1,13 @@
...
 using System.Collections;
 using System.Collections.Generic;
+using System.Linq;
 using System.Reflection;
+using JsonApiDotNetCore.Models;
 
 namespace JsonApiDotNetCore.Internal
 {
-    public static class TypeHelper
+    internal static class TypeHelper
```
Comment (wisepotato) [inline src/JsonApiDotNetCore/Internal/TypeHelper.cs]: is this not needed for unti tests??

File: test/UnitTests/Data/DefaultEntityRepository_Tests.cs [hunk 3]
```diff
@@ -77,14 +77,15 @@ public async Task UpdateAsync_Updates_Attributes_In_AttributesToUpdate()
 
         private DefaultEntityRepository<TodoItem> GetRepository()
         {
+
```
Comment (wisepotato) [inline test/UnitTests/Data/DefaultEntityRepository_Tests.cs]: empty spaaace

File: test/UnitTests/ResourceHooks/DiscoveryTests.cs [hunk 4]
```diff
@@ -14,7 +14,7 @@ public class DummyResourceDefinition : ResourceDefinition<Dummy>
         {
             public DummyResourceDefinition() : base(new ResourceGraphBuilder().AddResource<Dummy>().Build()) { }
```
Comment (wisepotato) [inline test/UnitTests/ResourceHooks/DiscoveryTests.cs]: space not needed here

File: src/JsonApiDotNetCore/Internal/TypeHelper.cs (hunk 2 above)
Comment (maurei) [reply src/JsonApiDotNetCore/Internal/TypeHelper.cs]: Nopes!

File: src/Examples/JsonApiDotNetCoreExample/Controllers/TodoCollectionsController.cs [hunk 5]
```diff
@@ -33,7 +33,7 @@ public override async Task<IActionResult> PatchAsync(Guid id, [FromBody] TodoIte
             if (entity.Name == "PRE-ATTACH-TEST")
             {
                 var targetTodoId = entity.TodoItems.First().Id;
-                var todoItemContext = _dbResolver.GetDbSet<TodoItem>();
+                var todoItemContext = _dbResolver.GetContext().Set<TodoItem>();
```
[the remark this answers is not in the document]
Comment (maurei) [reply src/Examples/JsonApiDotNetCoreExample/Controllers/TodoCollectionsController.cs]: Now it uses a EF core feature which wasn't there yet when JADNC introduced its own extension method to achieve the same  (`GetDbSet<T>`)

**Pull request merged**